halo/evmstaking2/keeper/keeper_internal_test.go

```diff
@@ -84,7 +90,59 @@ func TestInsertEVMEvents(t *testing.T) {
 	}
 }
 
-func setupKeeper(t *testing.T, submissionDelay int64) (*Keeper, sdk.Context) {
+func TestDeliverDelegate(t *testing.T) {
+	t.Parallel()
+
+	submissionDelay := int64(3)
+	ethClientMock, err := ethclient.NewEngineMock(
+		ethclient.WithPortalRegister(netconf.SimnetNetwork()),
+		ethclient.WithMockSelfDelegation(k1.GenPrivKey().PubKey(), 1),
+	)
+	require.NoError(t, err)
+
+	msgServer := &msgServerMock{}
+
+	keeper, ctx := setupKeeper(t, submissionDelay, ethClientMock, authKeeperMock{}, bankKeeperMock{}, stakingKeeperMock{}, msgServer)
+
+	var hash common.Hash
+	events, err := keeper.Prepare(ctx, hash)
+	require.NoError(t, err)
+
+	require.Len(t, events, 1)
+
+	for _, event := range events {
+		err := keeper.Deliver(ctx, hash, event)
+		require.NoError(t, err)
+	}
+
+	// Make sure the event was persisted.
+	insertedID := uint64(1)
+	found, err := keeper.eventsTable.Has(ctx, insertedID)
+	require.NoError(t, err)
+	require.True(t, found)
+
+	ctx = ctx.WithBlockHeight(submissionDelay)
+	err = keeper.EndBlock(ctx)
+	require.NoError(t, err)
+
+	// Make sure the event was deleted.
+	found, err = keeper.eventsTable.Has(ctx, insertedID)
+	require.NoError(t, err)
+	require.False(t, found)
+
+	// Assert that the message was delivered to the msg server.
+	require.Len(t, msgServer.delegateMsgBuffer, 1)
```

Comment:
we also need to ensure the actual fields of the thing is all as expected. To avoid bugs that mangle fields.